### PR TITLE
[Facebook Adapter] Initial corrections

### PR DESCRIPTION
**Pull request by matiasroldan6**
## Description
Did corrections according to previous work in past adapters to what was already done in Facebook Adapter.

- Remove unreferenced methods 
- Change type initializations to var 
- Add IDisposable handling 
- Remove this keyword
- Fix some Roslyn warnings

### CODE REVIEW

File: libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAPI.cs [hunk 1]
```diff
@@ -37,25 +38,30 @@ public FacebookAPI(string token, string secret, string apiHost = "graph.facebook
...
             }
 
             // send the request
-            HttpRequestMessage request = new HttpRequestMessage();
-            request.RequestUri = new Uri("https://" + this.apiHost + "/" + this.apiVersion + path + "?access_token=" + this.token + "&appsecret_proof=" + proof);
-            request.Method = method;
-            /* content type json? */
+            using (var request = new HttpRequestMessage())
+            {
+                request.RequestUri = new Uri("https://" + _apiHost + "/" + _apiVersion + path + "?access_token=" + _token + "&appsecret_proof=" + proof);
```
Comment (VictorGrycuk) [inline libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAPI.cs]: Why not use string interpolation?
`new Uri($"https://{ _apiHost }/{ _apiVersion + path + }?access_token={ _token }&appsecret_proof=" + proof);`

File: libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs [hunk 2]
```diff
@@ -241,16 +202,16 @@ public async Task ProcessAsync(HttpRequest request, HttpResponse response, IBot
...
                             }
                         }
                     }
 
                     // send code 200
                     response.StatusCode = 200;
                     response.ContentType = "text/plain";
-                    string text = string.Empty;
-                    await response.WriteAsync(text);
+                    var text = string.Empty;
```
Comment (VictorGrycuk) [inline libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs]: Why initialize an empty string? Could be simplified as `await response.WriteAsync(string.Empty).ConfigureAwait(false);` if it is not going to be used afterwards

File: libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs [hunk 3]
```diff
@@ -359,22 +321,27 @@ private async Task ProcessSingleMessageAsync(FacebookMessage message, BotCallbac
...
+            }
 
             if (expected != calculated)
             {
                 response.StatusCode = 401;
                 response.ContentType = "text/plain";
-                string text = string.Empty;
-                await response.WriteAsync(text);
+                var text = string.Empty;
+                await response.WriteAsync(text, cancellationToken).ConfigureAwait(false);
```
Comment (VictorGrycuk) [inline libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs]: Same comment about using `string.Empty` directly on this method

File: libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs [hunk 4]
```diff
@@ -4,70 +4,38 @@
...
         {
-            this.options = options;
-            this.options.ApiHost = "graph.facebook.com";
-            this.options.ApiVersion = "v3.2";
+            _options = options;
+            _options.ApiHost = "graph.facebook.com";
+            _options.ApiVersion = "v3.2";
 
-            if (string.IsNullOrEmpty(this.options.AccessToken) && this.options.GetAccessTokenForPageAsync != default(Func<string, Task<string>>))
+            if (string.IsNullOrEmpty(_options.AccessToken) && _options.GetAccessTokenForPageAsync != default(Func<string, Task<string>>))
```
Comment (ceciliaavila) [inline libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs]: Use IsNullOrWhiteSpace instead.

File: libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs [hunk 5]
```diff
@@ -4,70 +4,38 @@
...
             {
                 throw new Exception("Adapter must receive either an access_token or a getAccessTokenForPage function.");
```
Comment (ceciliaavila) [inline libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs]: this should probably be a NullArgumentException

File: libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs [hunk 6]
```diff
@@ -4,70 +4,38 @@
...
             }
 
-            if (string.IsNullOrEmpty(this.options.AppSecret))
+            if (string.IsNullOrEmpty(_options.AppSecret))
             {
                 throw new Exception("Provide an app_secret in order to validate incoming webhooks and better secure api requests");
```
Comment (ceciliaavila) [inline libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs]: Same here

File: libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs [hunk 7]
```diff
@@ -359,22 +321,27 @@ private async Task ProcessSingleMessageAsync(FacebookMessage message, BotCallbac
...
+        private async Task<bool> VerifySignatureAsync(HttpRequest request, HttpResponse response, CancellationToken cancellationToken)
         {
             var expected = request.Headers["x-hub-signature"];
-            var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(this.options.AppSecret));
-
-            var bodyStream = new StreamReader(request.Body);
 
-            var calculated = "sha1=" + hmac.ComputeHash(Encoding.UTF8.GetBytes(bodyStream.ReadToEnd()));
+            string calculated = null;
+            using (var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(_options.AppSecret)))
```
Comment (VictorGrycuk) [inline libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs]: I remember that this encryption was used multiple times on different botkits, wouldnt be useful to create a static class to do this? So one could just do `calculated = EncryptHelper.ToSHA256(_options.AppSecret, request.Body)`

File: libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs [hunk 8]
```diff
@@ -4,70 +4,38 @@
...
 {
     public class FacebookAdapter : BotAdapter
     {
         public const string NAME = "Facebook Adapter";
 
-        private readonly IFacebookAdapterOptions options;
+        private readonly FacebookAdapterOptions _options;
 
-        public FacebookAdapter(IFacebookAdapterOptions options)
+        public FacebookAdapter(FacebookAdapterOptions options)
```
Comment (ceciliaavila) [inline libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs]: The constructor should receive an instance of FacebookApi class.

File: libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs [hunk 9]
```diff
@@ -91,14 +59,14 @@ public async Task<FacebookAPI> GetAPIAsync(Activity activity)
                         pageId = activity.From.Id;
                     }
 
-                    string token = await this.options.GetAccessTokenForPageAsync(pageId);
+                    string token = await _options.GetAccessTokenForPageAsync(pageId).ConfigureAwait(false);
 
                     if (string.IsNullOrEmpty(token))
```
Comment (ceciliaavila) [inline libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs]: Replace all IsNullOrEmpty with IsNullOrWhiteSpace please.

File: libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs [hunk 10]
```diff
@@ -163,7 +124,7 @@ public override async Task<ResourceResponse[]> SendActivitiesAsync(ITurnContext
         public override async Task<ResourceResponse> UpdateActivityAsync(ITurnContext turnContext, Activity activity, CancellationToken cancellationToken)
         {
             // Facebook adapter does not support updateActivity.
-            return await Task.FromException<ResourceResponse>(new NotImplementedException("Facebook adapter does not support updateActivity."));
+            return await Task.FromException<ResourceResponse>(new NotImplementedException("Facebook adapter does not support updateActivity.")).ConfigureAwait(false);
```
Comment (ceciliaavila) [inline libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs]: Remove the async qualifier from this method and normalize the exception.

File: libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs [hunk 11]
```diff
@@ -176,7 +137,7 @@ public override async Task<ResourceResponse> UpdateActivityAsync(ITurnContext tu
         public override async Task DeleteActivityAsync(ITurnContext turnContext, ConversationReference reference, CancellationToken cancellationToken)
         {
             // Facebook adapter does not support deleteActivity.
-            await Task.FromException<ResourceResponse>(new NotImplementedException("Facebook adapter does not support deleteActivity."));
+            await Task.FromException<ResourceResponse>(new NotImplementedException("Facebook adapter does not support deleteActivity.")).ConfigureAwait(false);
```
Comment (ceciliaavila) [inline libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs]: Remove the async qualifier from this method and normalize the exception.

File: libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/Message.cs [hunk 12]
```diff
@@ -31,6 +29,6 @@ public class Message
         /// Gets or sets the quick replies.
         /// </summary>
         /// <value>The quick replies array.</value>
-        public List<object> QuickReplies { get; set; }
+        public List<object> QuickReplies { get; set; } = new List<object>();
```
Comment (ceciliaavila) [inline libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/Message.cs]: Is the set here necessary? If not, remove it and leave just the new List<().

Comment (matiasroldan6) [reply libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/Message.cs]: It is necessary. It's used in FacebookAdapter

File: libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs (hunk 7 above)
Comment (matiasroldan6) [reply libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs]: maybe

File: libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAPI.cs (hunk 1 above)
Comment (GasparAcevedoZainSouthworks) [inline libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAPI.cs]: Can we use String [formating](https://docs.microsoft.com/en-us/dotnet/api/system.string.format?view=netframework-4.8) or [interpolation](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/interpolated) here to make it more readable?

File: libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs (hunk 4 above)
Comment (GasparAcevedoZainSouthworks) [inline libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs]: Same as always on string.IsNullOrEmpty

File: libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs [hunk 13]
```diff
@@ -4,70 +4,38 @@
...
+            _options.ApiVersion = "v3.2";
 
-            if (string.IsNullOrEmpty(this.options.AccessToken) && this.options.GetAccessTokenForPageAsync != default(Func<string, Task<string>>))
+            if (string.IsNullOrEmpty(_options.AccessToken) && _options.GetAccessTokenForPageAsync != default(Func<string, Task<string>>))
             {
                 throw new Exception("Adapter must receive either an access_token or a getAccessTokenForPage function.");
             }
 
-            if (string.IsNullOrEmpty(this.options.AppSecret))
+            if (string.IsNullOrEmpty(_options.AppSecret))
```
Comment (GasparAcevedoZainSouthworks) [inline libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs]: Again about string.IsNullOrEmpty

File: libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs [hunk 14]
```diff
@@ -76,9 +44,9 @@ public async Task Init(Botkit botkit)
         /// <returns>A Facebook API client.</returns>
         public async Task<FacebookAPI> GetAPIAsync(Activity activity)
         {
-            if (!string.IsNullOrEmpty(this.options.AccessToken))
+            if (!string.IsNullOrEmpty(_options.AccessToken))
```
Comment (GasparAcevedoZainSouthworks) [inline libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs]: Same here

File: libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs [hunk 15]
```diff
@@ -359,22 +321,27 @@ private async Task ProcessSingleMessageAsync(FacebookMessage message, BotCallbac
...
-
-            var bodyStream = new StreamReader(request.Body);
 
-            var calculated = "sha1=" + hmac.ComputeHash(Encoding.UTF8.GetBytes(bodyStream.ReadToEnd()));
+            string calculated = null;
+            using (var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(_options.AppSecret)))
+            {
+                using (var bodyStream = new StreamReader(request.Body))
+                {
+                    calculated = "sha1=" + hmac.ComputeHash(Encoding.UTF8.GetBytes(bodyStream.ReadToEnd()));
```
Comment (GasparAcevedoZainSouthworks) [inline libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs]: Can we use here also string formating/interpolation?

File: libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs (hunk 4 above)
Comment (GasparAcevedoZainSouthworks) [inline libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs]: Also, here we are validation that options has an AccessToken, but we're not checking first if options is null or not. Add a validation which throws an ArgumentNullException for options (and for the Facabook API parameter after you add it, as @ceciliaavila mentioned above)

File: libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs (hunk 5 above)
Comment (GasparAcevedoZainSouthworks) [inline libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs]: Replace this exception with `ArgumentException`

File: libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs (hunk 6 above)
Comment (GasparAcevedoZainSouthworks) [inline libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs]: Replace also here with `ArgumentException`

File: libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs (hunk 9 above)
Comment (GasparAcevedoZainSouthworks) [inline libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs]: WhiteSpace

File: libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAPI.cs (hunk 1 above)
Comment (matiasroldan6) [reply libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAPI.cs]: isn't this the same that Victor said above?

File: libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs (hunk 4 above)
Comment (matiasroldan6) [reply libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs]: Isn't this the same that Ceci said above?

File: libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs (hunk 5 above)
Comment (matiasroldan6) [reply libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs]: Isn't this the same that Ceci said above?

File: libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs [hunk 16]
```diff
@@ -210,8 +218,7 @@ public async Task ProcessAsync(HttpRequest request, HttpResponse response, IBot
                     // send code 200
                     response.StatusCode = 200;
                     response.ContentType = "text/plain";
-                    var text = string.Empty;
-                    await response.WriteAsync(text).ConfigureAwait(false);
+                    await response.WriteAsync(string.Empty).ConfigureAwait(false);
```
Comment (ceciliaavila) [inline libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookAdapter.cs]: Here is missing the CancellationToken.

**Pull request merged**